crd/crd.go

```diff
@@ -0,0 +1,167 @@
+package crd
+
+import (
+	"path/filepath"
+
+	"github.com/giantswarm/microerror"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Config represents the configuration used to create a new CRD.
+type Config struct {
+	// Settings.
+	Group    string
+	Kind     string
+	Name     string
+	Plural   string
+	Singular string
+	Scope    string
+	Version  string
+}
+
+// DefaultConfig provides a default configuration to create a new CRD by best
+// effort.
+func DefaultConfig() Config {
+	return Config{
+		// Settings.
+		Group:    "",
+		Kind:     "",
+		Name:     "",
+		Plural:   "",
+		Singular: "",
+		Scope:    "",
+		Version:  "",
+	}
+}
+
+type CRD struct {
+	// Settings.
+	group    string
+	kind     string
+	name     string
+	plural   string
+	singular string
+	scope    string
+	version  string
+}
+
+// New creates a new CRD.
+func New(config Config) (*CRD, error) {
+	// Settings.
+	if config.Group == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Group must not be empty")
+	}
+	if config.Kind == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Kind must not be empty")
+	}
+	if config.Name == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Name must not be empty")
+	}
+	if config.Plural == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Plural must not be empty")
+	}
+	if config.Singular == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Singular must not be empty")
+	}
+	if config.Scope == "" {
```

Comment:
I'd dump this check, the next one handles `config.Scope` not being valid, and with a better error message. No blocker.

Reply:
I want to keep it because the error case has a different root cause. Scope must not be empty and when you set it it has to match a specific set of valid options.